### PR TITLE
Add Terminal Progress

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    * This function flushes the output and should not be called too often.
+    */
+  private def print(phase: String, msg: String): Unit = synchronized {
+    // Compute the next character in the spinner.
+    tick = (tick + 1) % SpinnerChars.length
+    val spinner = SpinnerChars(tick)
+
+    // Build the string to print and pad it.
+    val s = s"$spinner [$phase] $msg"
+    val r = s + " " * (80 - s.length)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/util/ProgressBar.scala:65]: I think there's a String.pad function

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Print the string followed by carriage return.
+    System.out.print(s"$r\r")
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/util/ProgressBar.scala:68]: Why not println?

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+class ProgressBar {
+  /**
+    * The characters in the spinner.
+    */
+  private val SpinnerChars = Array("|", "/", "-", "\\")
+
+  /**
+    * The sample rate.
+    */
+  private val SampleRate: Int = 100
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/util/ProgressBar.scala:27]: What's the reason for doing an update per X ticks rather than per X time?

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:65]: Ill try it!

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:68]: Because I am not printing newlines :) I would upload a GIF, if I knew how to make one.

TLDR: I am actually overwriting the same line again and again.

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:27]: Here is the overall concern: I want this to be fast. It should not slow down the compiler.

Problem is: I have to call flush (which slows it down) and I am invoked from multiple threads (so I should try to be thread safe). This leads to the very awkward code that is there.

Oh, and as for ticks per time, I assume is too expensive to call `nanoTime` so frequently.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:27]: I am open to better ideas of course.

**Comment by magnus-madsen** [comment]
Any suggestions for how to tweak the output?

File: main/src/ca/uwaterloo/flix/util/ProgressBar.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:29]: We could combine them, checking time every X ticks and updating if at least Y time has passed. But maybe that's more complex than it's worth.

One thing that would be nice to have is if one particular function takes a long time to process, it would be shown for a long time on screen (matching user's intuition). That's not the case with either of these implementations.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:29]: I agree, but only in principle. What you propose is to give some useful information (but that would only be useful to us). 

What I propose is to "show the programmer that something is happening".

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:29]: I am also abusing the same trick as npm and many other tools in that I appear to give the idea that the compiler is working fast. You see lots of text flash by. If I only showed "Typer" then it would appear slow. 

In other words, the purpose of this particular exercise is to give the human emotion of "Ah the compiler is running. Great! And its fast. Something is happening!".

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:29]: Airplanes don't have to be pained, but who would want to fly with a plane where the coating is missing or falling of? Its all about appearances.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ProgressBar.scala:29]: which is why we should probably add some color and more (useless) information.
